documentation for l3afd config options (#86)
l3af-project/l3afd: docs/configdoc.md

|kernel-minor-version|`"15"`|Minor version of the kernel (Ex 4.15)| Only on linux |
|shutdown-timeout|`"1s"`|It is the amount of time allowed for L3afd graceful stop ( stopping l3afd).After shutdown-timeout L3afd itself exit| No |
|http-client-timeout|`"10s"`|Maximum amount of time allowed to get HTTP response headers when fetching a program from a repository| No |
|max-nf-restart-count|`"3"`|Maximum number of try's to restart eBPF programs if they are not running| No |
|bpf-chaining-enabled|`"true"`|Boolean to set bpf-chaining. For more info about bpf chaining check [L3AF_KFaaS.pdf](https://github.com/l3af-project/l3af-arch/blob/main/L3AF_KFaaS.pdf)| Yes |
|swagger-api-enabled|`"false"`|Whether the swagger API is enabled or not.  For more info see [swagger.md](https://github.com/l3af-project/l3afd/blob/main/docs/swagger.md)| No |
|environment|`"PROD"`|If set to anything other than "PROD", mTLS security will not be checked| Yes |
|BpfMapDefaultPath|`"/sys/fs/bpf"`|The base pin path for eBPF maps| Yes |

## [kf-repo]
| FieldName     | Example       | Description     | Required        |
| ------------- | ------------- | --------------- | --------------- |

> |url| `"http://localhost:8000/"`|Default repository from which to download eBPF programs| No |

Reviewer: ```suggestion
|url| `"http://localhost:8000/"`|Default repository from which to download eBPF packages| No |
```

## [web]
| FieldName     | Example       | Description     |  Required        |
| ------------- | ------------- | --------------- |  --------------- |
|metrics-addr|`"0.0.0.0:8898"`|Prometheus endpoint for pulling/scraping the metrics.  For more info about Prometheus see [prometheus.io](https://prometheus.io/) | Yes |
|kf-poll-interval|`"30s"`|Periodic interval at which to scrape metrics using Prometheus| Yes |
|n-metric-samples|`"20"`|Number of Metric Samples| Yes |

## [xdp-root-program]
| FieldName     | Example       | Description     | Required        |
| ------------- | ------------- | --------------- | --------------- |
|name|`"xdp-root"`|Name of xdp-root program| Yes |

> |artifact|`"xdp-root.tar.gz"`|Archive of xdp-root program. Only tar.gz and .zip formats are supported| Yes |

Reviewer: ```suggestion
|artifact|`"xdp-root.tar.gz"`|Filename of xdp-root package. Only tar.gz and .zip formats are supported| Yes |
```

|ingress-map-name|`"root_array"`|Ingress map name of xdp-root program| Yes |
|command|`"xdp_root"`|Command to run xdp-root program| Yes |
|version|`"1.01"`|Version of xdp-root program| Yes |

> |user-program-daemon|`"false"`|Boolean to check xdp-root is user-program daemon or not| Yes |

Reviewer: Explain what this means in more detail.  I think it means if false, the program will exit when started, and if true, then it will run until stopped by l3afd.  Correct?
```suggestion
|user-program-daemon|`"false"`|Set to true it requires l3afd to stop the application (via SIGTERM on Linux or SIGKILL on Windows)| Yes |
```

## [tc-root-program]
| FieldName     | Example       | Description     | Required        |
| ------------- | ------------- | --------------- | --------------- |

> |name|`"tc_root"`|Name of tc_root program| Yes |
> |artifact|`"l3af_tc_root.tar.gz"`|Archive of tc_root program| Yes |

Reviewer: ```suggestion
|name|`"tc_root"`|Name of subdirectory in which to extract artifact| Yes |
|artifact|`"l3af_tc_root.tar.gz"`|Filename of tc_root package| Yes |
```

|ingress-map-name|`"tc_ingress_root_array"`|Ingress map name of tc_root program| Yes |
|egress-map-name|`"tc_egress_root_array"`|Egress map name of tc_root program,for more info about ingress/egress check [cilium](https://docs.cilium.io/en/v1.9/concepts/ebpf/intro/)| Yes |
|command|`"tc_root"`|Command to run tc_root program| Yes |
|version|`"1.0"`|Version of tc_root program| Yes |
|user-program-daemon|`"false"`|Boolean to check xdp-root is user-program daemon or not| Yes |

## [l3af-configs]
| FieldName     | Example       | Description     | Required        |
| ------------- | ------------- | --------------- | --------------- |

> |restapi-addr|`"localhost:53000"`| Hostname and Port of l3af-configs rest api | Yes |

Reviewer: ```suggestion
|restapi-addr|`"localhost:53000"`| Hostname and Port of l3af-configs REST API | Yes |
```

# [l3af-config-store]
| FieldName     | Example       | Description     | Required        |
| ------------- | ------------- | --------------- | --------------- |
|filename|`"/etc/l3afd/l3af-config.json"`|Absolute path of persistent config file where we are storing L3afBPFPrograms objects. For more info see [models](https://github.com/l3af-project/l3afd/blob/main/models/l3afd.go)| Yes |

# [mtls]
| FieldName     | Example       | Description     | Required        |
| ------------- | ------------- | --------------- | --------------- |
|enabled| `"true"` | Boolean to check mtls enabled or not on REST API exposed by l3afd| Yes |
|min-tls-version|`"1.3"`| Minimum tls version allowed| No |
|cert-dir|`"/etc/l3af/certs"`|Absolute path of ca certificates. In Linux Context it is pointing to a filesystem directory, but in Windows it can point to [certificate store](https://docs.microsoft.com/en-us/windows-hardware/drivers/install/certificate-stores) | Yes |
|server-crt-filename|`"server.crt"`|Server's ca certificate filename| Yes |
|server-key-filename|`"server.key"`|Server's mtls key filename| Yes |
|cert-expiry-warning-days|`"30"`|How many days before expiry you want warning| No |

> 
> 
> 
> 
> 

Reviewer: ```suggestion
```

# L3AFD Config Options Documentation

See [l3afd.cfg](https://github.com/l3af-project/l3af-arch/blob/main/dev_environment/cfg/l3afd.cfg) for a full example configuration.

```
[DEFAULT]

[l3afd]
pid-file: ./l3afd.pid
datacenter: dummy
bpf-dir: /dev/shm
bpf-log-dir:
kernel-major-version: 4
kernel-minor-version: 15
shutdown-timeout: 1s
http-client-timeout: 10s
max-nf-restart-count: 3
bpf-chaining-enabled: true
swagger-api-enabled: false
# PROD | DEV
environment: PROD
....
```

### Below is the detailed documentation for each field

## [l3afd]
| FieldName     | Example       | Description     | Required        |
| ------------- | ------------- | --------------- | --------------- |

> |pid-file| `"./l3afd.pid"`  | The pin Path for l3afd.pid file which contains process id of eBPF Programs | Yes |

Reviewer: Is it really the process id of ebpf programs (which should be executing in the kernel process, not in l3afd) or is it really the process id of l3afd itself?

Author: yes it is storing pid of l3afd.

|datacenter| `"dummy"` | Name of Datacenter| Yes |

> |bpf-dir| `"/dev/shm"` | Absolute Path of eBPF Programs (where they are extracted) | Yes |

Reviewer: Question: is it only the ebpf program itself, or also the application that may be in the package?  If the latter, then the description here is wrong.

Author: The Extracted directory contain only ebpf program binary and ebpf bytecode.
There is no package

> |bpf-log-dir|`""`      | Absolute Path, where we store logs of eBPF Programs.We are not storing logs of l3afd actions on eBPF Programs| No |

Reviewer: ```suggestion
|bpf-log-dir|`""`      | Absolute Path, where we store logs of eBPF Programs. We are not storing logs of l3afd actions on eBPF Programs| No |
```

Author: done

> |kernel-major-version|`"4"`|Major version of the kernel| Yes(Only for linux) |
> |kernel-minor-version|`"15"`|Minor version of the kernel (Ex 4.15)| Yes(Only for linux) |

Reviewer: ```suggestion
|kernel-major-version|`"4"`|Major version of the kernel| Only on linux |
|kernel-minor-version|`"15"`|Minor version of the kernel (Ex 4.15)| Only on linux |
```

> |shutdown-timeout|`"1s"`|It is the amount of time allowed for L3afd graceful stop ( stopping l3afd).If L3afd graceful stop is not successful then os interrupt stop the l3afd| No |

Reviewer: I find this description confusing as it sounds like this is configuring the OS.  Glancing at the code, it looks like this is configuring l3afd itself to exit after that time, rather than causing the OS to kill the l3afd process.

Author: I have changed.

> |http-client-timeout|`"10s"`|It is the amount of time allowed to get response headers to fetch a program from a repository via HTTP| No |

Reviewer: ```suggestion
|http-client-timeout|`"10s"`|Maximum amount of time allowed to get HTTP response headers when fetching a program from a repository| No |
```

Author: done

|max-nf-restart-count|`"3"`|Maximum number of try's to restart eBPF programs if they are not running| No |
|bpf-chaining-enabled|`"true"`|Boolean to set bpf-chaining. For more info about bpf chaining check [L3AF_KFaaS.pdf](https://github.com/l3af-project/l3af-arch/blob/main/L3AF_KFaaS.pdf)| Yes |

> |swagger-api-enabled|`"false"`|Boolean to check swagger api enable or not, for more info check [swagger.md](https://github.com/l3af-project/l3afd/blob/main/docs/swagger.md)| No |

Reviewer: ```suggestion
|swagger-api-enabled|`"false"`|Whether the swagger API is enabled or not.  For more info see [swagger.md](https://github.com/l3af-project/l3afd/blob/main/docs/swagger.md)| No |
```

Author: done

> |environment|`"PROD"`| Environment  variable for l3afd, for DEV environment we will not check mtls security | Yes |

Reviewer: "Environment variable" sounds like a Linux or Windows environment variable.
```suggestion
|environment|`"PROD"`| If set to anything other than "PROD", mTLS security will not be checked.  | Yes |
```
The description is incorrect if it implies that only "DEV" will not check mtls security

Author: done

|BpfMapDefaultPath|`"/sys/fs/bpf"`|The base pin path for eBPF maps| Yes |

> |TcMapsRelativePath|`"tc/globals"`|The relative pin path of tc maps with respect to BpfMapDefaultPath.We will store xdp maps in BpfMapDefaultPath | Yes |

Reviewer: ```suggestion
|TcMapsRelativePath|`"tc/globals"`|The relative pin path of tc maps with respect to BpfMapDefaultPath.  L3afd currently stores all other maps directly in BpfMapDefaultPath | Yes |
```

Author: done

## [kf-repo]
| FieldName     | Example       | Description     | Required        |
| ------------- | ------------- | --------------- | --------------- |
|url| `"http://localhost:8000/"`|Default repository from which to download eBPF programs| No |

## [web]
| FieldName     | Example       | Description     |  Required        |
| ------------- | ------------- | --------------- |  --------------- |

> |metrics-addr|`"0.0.0.0:8898"`|Prometheus endpoint for pull/scrape the metrics, for more info about Prometheus check [prometheus.io](https://prometheus.io/) | Yes |

Reviewer: ```suggestion
|metrics-addr|`"0.0.0.0:8898"`|Prometheus endpoint for pulling/scraping the metrics.  For more info about Prometheus see [prometheus.io](https://prometheus.io/) | Yes |
```
Grammar

Author: done

> |kf-poll-interval|`"30s"`|Periodic Interval to scrape metrics using Prometheus| Yes |

Reviewer: ```suggestion
|kf-poll-interval|`"30s"`|Periodic interval at which to scrape metrics using Prometheus| Yes |
```
grammar

Author: done

|n-metric-samples|`"20"`|Number of Metric Samples| Yes |

## [xdp-root-program]
| FieldName     | Example       | Description     | Required        |
| ------------- | ------------- | --------------- | --------------- |

> |name|`"xdp-root"`|Name of xdp-root program| Yes |

Reviewer: Is this used only for display/logging purposes (so it doesn't have to match anything anywhere else) or does the name need to match something inside the artifact?

Author: Basically we used program name in  extracted folder name.
You can see in GetArtifacts function.

Reviewer: ```suggestion
|name|`"xdp-root"`|Name of subdirectory in which to extract artifact| Yes |
```

## [tc-root-program]
| FieldName     | Example       | Description     | Required        |
| ------------- | ------------- | --------------- | --------------- |
|name|`"tc_root"`|Name of tc_root program| Yes |
|artifact|`"l3af_tc_root.tar.gz"`|Archive of tc_root program| Yes |
|ingress-map-name|`"tc_ingress_root_array"`|Ingress map name of tc_root program| Yes |
|egress-map-name|`"tc_egress_root_array"`|Egress map name of tc_root program,for more info about ingress/egress check [cilium](https://docs.cilium.io/en/v1.9/concepts/ebpf/intro/)| Yes |
|command|`"tc_root"`|Command to run tc_root program| Yes |
|version|`"1.0"`|Version of tc_root program| Yes |
|user-program-daemon|`"false"`|Boolean to check xdp-root is user-program daemon or not| Yes |

## [l3af-configs]
| FieldName     | Example       | Description     | Required        |
| ------------- | ------------- | --------------- | --------------- |
|restapi-addr|`"localhost:53000"`| Hostname and Port of l3af-configs rest api | Yes |

# [l3af-config-store]
| FieldName     | Example       | Description     | Required        |
| ------------- | ------------- | --------------- | --------------- |

> |filename|`"/etc/l3afd/l3af-config.json"`|Absolute path of persistent config file where we are storing L3afBPFPrograms object's , for more info check [models](https://github.com/l3af-project/l3afd/blob/main/models/l3afd.go)| Yes |

Reviewer: ```suggestion
|filename|`"/etc/l3afd/l3af-config.json"`|Absolute path of persistent config file where we are storing L3afBPFPrograms objects.  For more info see [models](https://github.com/l3af-project/l3afd/blob/main/models/l3afd.go)| Yes |
```

Author: done

# [mtls]
| FieldName     | Example       | Description     | Required        |
| ------------- | ------------- | --------------- | --------------- |
|enabled| `"true"` | Boolean to check mtls enabled or not on REST API exposed by l3afd| Yes |
|min-tls-version|`"1.3"`| Minimum tls version allowed| No |

> |cert-dir|`"/etc/l3af/certs"`|Absolute path of ca certificates. In Windows we store ca certificates in [certificate store](https://docs.microsoft.com/en-us/windows-hardware/drivers/install/certificate-stores),but in l3afd context we will store in filesystem.This is the limitation of l3afd.  | Yes |

Reviewer: ```suggestion
|cert-dir|`"/etc/l3af/certs"`|Absolute path of ca certificates. In Windows we store ca certificates in [certificate store](https://docs.microsoft.com/en-us/windows-hardware/drivers/install/certificate-stores),but in l3afd context we will store in filesystem.This is the limitation of l3afd.  | Yes |
```
This text seems contradictory.  It implies that on windows l3afd stores certs in the certificate store, but in a l3afd windows context, they're not.   That sounds like a contradiction.  Please reword.
